### PR TITLE
[Sparse] Support broadcasting operators

### LINKED ISSUE #5370: [Sparse] Support broadcasting SparseMatrix <op> 1-D tensor. <op> can be +, -, *, and /.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #5370 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/sparse/broadcast.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        v = v.view(-1)[A.col]
+    else:
+        v = v.view(-1)[A.row]
+    if A.val.dim() > 1:
+        v = v.view(-1, 1)
+    ret_val = op(A.val, v)
+    return val_like(A, ret_val)
+
+
+def broadcast_add(A: SparseMatrix, v: torch.Tensor) -> SparseMatrix:
```
Comment (jermainewang) [inline python/dgl/sparse/broadcast.py:100]: Are these operators for internal usage or user-facing? If the later, should add them to the API doc.

File: python/dgl/sparse/broadcast.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    )
+    assert v.dim() <= 2 and (1 in v.shape), shape_error_message
+    broadcast_dim = None
+    for d, (dim1, dim2) in enumerate(zip(A.shape, v.shape)):
+        assert dim2 in (1, dim1), shape_error_message
+        if dim1 != dim2:
+            assert broadcast_dim is None, shape_error_message
+            broadcast_dim = d
+
+    # A and v has the same shape of (1, *) or (*, 1)
```
Comment (frozenbugs) [inline python/dgl/sparse/broadcast.py:86]:
```suggestion
    # A and v has the same shape of (1, *) or (*, 1).
```

File: python/dgl/sparse/broadcast.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                 values=tensor([[ 8, 18],
+                                [29, 39],
+                                [47, 57]]),
+                 shape=(3, 4), nnz=3, val_size=(2,))
+    """
+    op = getattr(operator, op)
+    if v.dim() == 1:
+        v = v.view(1, -1)
+
+    shape_error_message =  (
```
Comment (frozenbugs) [inline python/dgl/sparse/broadcast.py:74]:
```suggestion
    shape_error_message = (
```

File: python/dgl/sparse/broadcast.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        f"Dimension mismatch for broadcasting. Got A.shape = {A.shape} and"
+        f"v.shape = {v.shape}."
+    )
+    assert v.dim() <= 2 and (1 in v.shape), shape_error_message
+    broadcast_dim = None
+    for d, (dim1, dim2) in enumerate(zip(A.shape, v.shape)):
```
Comment (frozenbugs) [inline python/dgl/sparse/broadcast.py:80]: Can you add a comment to explain the broadcast_dim calculation logic here?
It is not obvious why a for loop is needed,

File: python/dgl/sparse/broadcast.py (hunk 1 above)
Comment (frozenbugs) [reply python/dgl/sparse/broadcast.py:100]: +1, also should A + V support broadcast by default?

Comment (czkkkkkk) [reply python/dgl/sparse/broadcast.py:100]: We have discussed that broadcast ops are special so they should be isolated from operator +.

File: python/dgl/sparse/broadcast.py (hunk 4 above)
Comment (czkkkkkk) [reply python/dgl/sparse/broadcast.py:84]: Added.

File: python/dgl/sparse/broadcast.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/sparse/broadcast.py:104]: Added to documentation.

**Review by jermainewang — APPROVED** [review]

**Comment by jermainewang** [comment]
@dgl-bot

**Pull request merged**